### ISSUE TITLE
How does Dropout2d help in cloze task?

### ISSUE THREAD

**Issue by sai-prasanna** [huggingface/pytorch-openai-transformer-lm#11]
``` python
class ClfHead(nn.Module):
    """ Classifier Head for the transformer """

    def __init__(self, clf_token, cfg):
        super(ClfHead, self).__init__()
        self.n_embd = cfg.n_embd
        self.clf_token = clf_token
        self.dropout = nn.Dropout2d(cfg.clf_pdrop)  # To reproduce the noise_shape parameter of TF implementation
        self.linear = nn.Linear(cfg.n_embd, 1)
        nn.init.normal_(self.linear.weight, std=0.02)
        nn.init.normal_(self.linear.bias, 0)

    def forward(self, h, x):
        # Classification logits
        clf_h = h.view(-1, self.n_embd)
        flat = x[:, :, :, 0].contiguous().view(-1)
        clf_h = clf_h[flat == self.clf_token, :]
        clf_h = clf_h.view(-1, x.size(1), self.n_embd, 1)
        clf_h = self.dropout(clf_h)
        clf_h = clf_h.view(-1, self.n_embd)
        clf_logits = self.linear(clf_h)
        return clf_logits.view(-1, x.size(1))
```
Here the `self.dropout(clf_h)` essentially removes the representation of a sentence and its conclusion, there is remote chance (0.2*0.2) that both representations get removed for a given data item. I am confused on how this aids training .

**Comment by chu-tianxiang** [comment]
If I'm not mistaken, the input of Dropout2d is (N, C, H, W) where C corresponds to x.size(1), i.e. 2, so it will only force the same dimensions are dropped out for both sentences.

**Comment by sai-prasanna** [comment]
Pytorch docs for nn.Dropout2d
> Randomly zeroes whole channels of the input tensor. The channels to zero-out are randomized on every forward call.

I tried Dropout2d , it zeroes out the entire channel, here sentence.

**Comment by thomwolf** [comment]
Interesting, thanks @sai-prasanna for investigating that.
I had some interrogation on how to best reproduce the `noise_shape` parameters of the TF implementation indeed (the original is here: https://github.com/openai/finetune-transformer-lm/blob/master/train.py#L189).
`Dropout2d` might not be a good option. Do you have any idea on an alternative option?
Otherwise, maybe a simple `Dropout. I will try to benchmark the two options.

**Comment by sai-prasanna** [comment]
Instead of nn.Dropout we can create a mask using torch.bernoulli and apply it to both the sentences with broadcasted multiply? https://discuss.pytorch.org/t/how-to-fix-the-dropout-mask-for-different-batch

We have to make sure that scaling is done, and its not applied in eval mode .

**Comment by rodgzilla** [comment]
Ok so I ran an experiment on this section of the code. With the current version of the code (using nn.Dropout2d) I get the following output:

```
device cuda n_gpu 1
Encoding dataset...
Loading weights...                                                              
running epoch 0
Logging
1 187 2.277 7.512 94.92 88.50                                                   
running epoch 1
Logging                                                                         
2 374 0.593 11.446 99.47 88.50                                                  
running epoch 2
Logging                                                                         
3 561 0.315 14.687 99.73 87.97
```

I modified the code to use the classical dropout:

```python
class ClfHead(nn.Module):
    """ Classifier Head for the transformer """

    def __init__(self, clf_token, cfg):
        super(ClfHead, self).__init__()
        self.n_embd = cfg.n_embd
        self.clf_token = clf_token
        self.dropout = nn.Dropout(cfg.clf_pdrop)  # To reproduce the noise_shape parameter of TF implementation
        self.linear = nn.Linear(cfg.n_embd, 1)
        nn.init.normal_(self.linear.weight, std=0.02)
        nn.init.normal_(self.linear.bias, 0)

    def forward(self, h, x):
        clf_h = h.view(-1, self.n_embd)
        flat = x[..., 0].contiguous().view(-1)
        clf_h = clf_h[flat == self.clf_token, :]
        clf_h = self.dropout(clf_h)
        clf_logits = self.linear(clf_h)

        return clf_logits.view(-1, x.size(1))
```

I am getting the following output which seems significantly better. 

```
device cuda n_gpu 1
Encoding dataset...
Loading weights...                                                              
running epoch 0
Logging                                                                         
1 187 2.840 8.655 94.92 89.84                                                   
running epoch 1
Logging                                                                         
2 374 0.858 15.130 99.20 91.44                                                  
running epoch 2
Logging                                                                         
3 561 0.000 15.668 100.00 90.91
```

Is what I'm doing correct ? And if so, should I create a pull request ?

**Comment by BangLiu** [comment]
@rodgzilla I got different results.
When using nn.Dropout2D, I got

running epoch 0
	Logging                                                                         
	1 187 7.826 9.969 78.34 75.94                                                   
	running epoch 1
	Logging                                                                         
	2 374 1.318 6.060 97.59 90.11                                                   
	running epoch 2
	Logging                                                                         
	3 561 0.368 12.730 99.47 90.37                                                  
	../../../datasets/original/ROCStories/                                          
	checkpoints/OpenAITransformer/ROCStories.tsv
	checkpoints/OpenAITransformer/rocstories.jsonl
	ROCStories Valid Accuracy: 90.37                                                
	ROCStories Test Accuracy:  88.13

When using classical dropout, I got:

	running epoch 0
	Logging                                                                         
	1 187 2.907 10.442 95.99 88.50                                                  
	running epoch 1
	Logging                                                                         
	2 374 0.722 13.166 99.73 89.84                                                  
	running epoch 2
	Logging                                                                         
	3 561 0.002 15.808 100.00 89.84                                                 
	../../../datasets/original/ROCStories/                                          
	checkpoints/OpenAITransformer/ROCStories.tsv
	checkpoints/OpenAITransformer/rocstories.jsonl
	ROCStories Valid Accuracy: 89.84                                                
	ROCStories Test Accuracy:  86.26

Seems the performance is kind of random.
I use CPU.

**Comment by rodgzilla** [comment]
@BangLiu You're right, I did not run enough experiments. 

I launched the code with 5 different seeds (42, 43, 44, 45, 46) with classical dropout and I get the following validation accuracies after the third epoch: `90.01, 89.30, 91.44, 88.5, 90.37`.

I then launched the code with the same 5 seeds using the 2D version of dropout and I get the following result: `87.97, 89.84, 92.25, 86.9, 90.11`.

The result are marginally better with the classical dropout in 3 of the 5 cases but I don't think the sample size is big enough to draw any kind of conclusion. 

If I understand correctly the OpenAI code that @thomwolf linked above, dropout2d should be the correct way to implement their algorithm.

```python
clf_h = tf.reshape(clf_h, [-1, 2, n_embd])
if train and clf_pdrop > 0:
    shape = shape_list(clf_h)
    shape[1] = 1
    clf_h = tf.nn.dropout(clf_h, 1-clf_pdrop, shape)
```

the shape given to dropout should be `(batch_size, 1, n_embd)`. The TF documentation of `tf.nn.dropout` is saying this:

>By default, each element is kept or dropped independently. If noise_shape is specified, it must be broadcastable to the shape of x, and only dimensions with noise_shape[i] == shape(x)[i] will make independent decisions. For example, if shape(x) = [k, l, m, n] and noise_shape = [k, 1, 1, n], each batch and channel component will be kept independently and each row and column will be kept or not kept together."

As the code have `shape[1] = [1]` and dimension 1 is corresponding to the sentence selection index, I think dropout 2D corresponds to this behavior. As I'm not too familiar with TensorFlow could someone check what I am saying ?

Is it what they really wanted to do ? The research paper only states the following:

>*Fine-tuning details* Unless specified, we reuse the hyperparameter settings from unsupervised pre-training. We add dropout to the classifier with a rate of 0.1.

They do not precise that they drop all the network outputs corresponding to a specific sentence but at the same time they reshape `clf_h` specifically for this purpose. Any opinion ?

**Comment by sai-prasanna** [comment]
@rodgzilla  I don't know tensorflow that much. But From the mask's documentation I understood that, the mask will apply the same dropout mask for the sentence pair dimension.

In Open AI code we have shape as (batch_size, 1, n_embd)
Data size is (batch_size, 2, n_embed), so essentially for a given sentence pair , same  dropout is applied.

For given batch item dropout mask will be calculated independently for each dimension on n_embed dimension . Statistically p% of the activations will be zero. But only one such mask is calculated for the (sentece,left) and (sentence,right) pair.

eg,

Lets say batch_size is 2 and n_embed is 3 
```

activation = [
                        [                        
                           [ 1, 2, 3],
                           [ 3, 4, 5]
                         ],
                        [                        
                           [6, 7, 8],
                           [9, 10, 11]
                         ],
                       ]
```

After applying dropout of 0.5 (without scaling by 1/(1-p))

```
activation = [
                        [                        
                           [ 1,0, 3],            #Second dimension is dropped out in sentence pairs
                           [ 3, 0, 5]            
                         ],
                        [                        
                           [0,0, 8],              #First and second dimension is dropped out in sentence pairs
                           [0,0, 11]  
                         ],
                       ]
```


But Pytorch Dropout2D just drops out the entire (sentence,right) and (sentence,left) pair , zeroing out all the values. Same example after doing pytorch dropout2d,

```
activation = [
                        [                          #Channel not dropped out 
                           [ 1,2, 3],            
                           [ 3, 4, 5]            
                         ],
                        [                         # Channel dropped out
                           [0,0, 0],            
                           [0,0, 0]  
                         ],
                       ]
```
I think a  prior to do this would be to force the predictor to learn to distinguish which is correct answer from different sets of same dimensions.

**Comment by sai-prasanna** [comment]
@rodgzilla  To accomplish the same thing as tensorflow, we have to create a dropout mask using torch.rand, for each batch of n_embed dimension, Multiply the same mask to both the rows in dimension 1. Then multiply by 1/(1-p), And just do this in training time ( if self.training:) .

**Comment by rodgzilla** [comment]
@sai-prasanna I was not sure if you were right about the behavior of `tf.nn.dropout` when using the `noise_shape` argument so [I tried it](https://github.com/rodgzilla/notebook_sharing/blob/master/dropout_noise_shape.ipynb) and it turns out you were right.

I will work on the proper way to implement this behavior in PyTorch and will create a pull request.

**Comment by rodgzilla** [comment]
Ok, I think I have a very simple fix.

We can indeed use `torch.nn.Dropout2d` to reproduce this behavior, the only problem is that we do not want to drop a channel (every value corresponding to a sentence) but a "column" (a specific value for each of the sentences forming an input). So we just have to transpose the tensor before applying the dropout and transpose again right after (in a way we conjugate by a transposition).

This following example in a Python interpreter should be clearer.

```python
In [88]: t = torch.randn(3,2,4)

In [89]: t
Out[89]: 
tensor([[[ 0.7503,  0.7338, -0.3102,  0.3698],
         [ 0.6638, -0.9277,  0.1738, -0.4256]],

        [[-2.5322,  0.2958,  0.1383, -0.1261],
         [ 2.2745,  0.2616, -0.1744,  0.2710]],

        [[ 0.7897, -1.6991, -0.1027, -0.2370],
         [-0.8105, -0.2241, -0.1237, -1.5941]]])

In [90]: dropout = torch.nn.Dropout2d(0.5)

In [91]: dropout(t.transpose(1,2)).transpose(1,2)
Out[91]: 
tensor([[[ 1.5006,  1.4676, -0.0000,  0.0000],
         [ 1.3276, -1.8554,  0.0000, -0.0000]],

        [[-5.0645,  0.5917,  0.2766, -0.0000],
         [ 4.5490,  0.5233, -0.3488,  0.0000]],

        [[ 1.5794, -0.0000, -0.2054, -0.0000],
         [-1.6209, -0.0000, -0.2475, -0.0000]]])

```

I am on my laptop at the moment so I am not able to test the new version of the code properly (no GPU), I'll test it tomorrow morning and if it is fine I will create a pull request.

I think that PyTorch should natively allow to do this kind of things. I will try to implement it and create a pull request for it.

**Comment by rodgzilla** [comment]
I created the pull request for this fix. 

I ran a few trainings just to be sure that it is working and I get the following results:

Seed 42
```
device cuda n_gpu 1
Encoding dataset...
Loading weights...                                                              
running epoch 0
Logging                                                                         
1 187 4.358 7.397 91.18 84.49                                                   
running epoch 1
Logging                                                                         
2 374 0.807 8.412 99.20 90.37                                                   
running epoch 2
Logging                                                                         
3 561 0.000 20.528 100.00 90.11                                                 
ROCStories Valid Accuracy: 90.37                                                
ROCStories Test Accuracy:  87.17
```

Seed 43
```
device cuda n_gpu 1
Encoding dataset...
Loading weights...                                                              
running epoch 0
Logging                                                                         
1 187 1.390 8.253 96.52 89.30                                                   
running epoch 1
Logging                                                                         
2 374 0.098 13.438 99.73 90.91                                                  
running epoch 2
Logging                                                                         
3 561 0.000 16.577 100.00 91.18                                                 
ROCStories Valid Accuracy: 91.18                                                
ROCStories Test Accuracy:  87.17
```

Seed 44
```
device cuda n_gpu 1
Encoding dataset...
Loading weights...                                                              
running epoch 0
Logging                                                                         
1 187 3.236 7.552 91.18 83.69                                                   
running epoch 1
Logging                                                                         
2 374 1.036 12.025 98.66 86.36                                                  
running epoch 2
Logging                                                                         
3 561 0.055 17.220 99.73 86.90                                                  
ROCStories Valid Accuracy: 86.90                                                
ROCStories Test Accuracy:  84.66
```
